File: source/ch/cyberduck/core/sparkle/Sandbox.m
/*
 *  Copyright (c) 2014 David Kocher. All rights reserved.
 *  http://cyberduck.io/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.io
 */

#import <Sandbox.h>
#import <Cocoa/Cocoa.h>
#import <Security/SecRequirement.h>
#import <Security/SecStaticCode.h>
#import <objc/runtime.h>

JNIEXPORT jboolean JNICALL Java_ch_cyberduck_core_sparkle_Sandbox_isSandboxed(JNIEnv *env, jobject this)
{
    BOOL isSandboxed = NO;
    NSURL *bundleURL = [[NSBundle mainBundle] bundleURL];
    SecStaticCodeRef staticCode = NULL;
    SecStaticCodeCreateWithPath((CFURLRef)bundleURL, kSecCSDefaultFlags, &staticCode);
    if (!staticCode) {
        return NO;
    }
    static SecRequirementRef sandboxRequirement = NULL;
    SecRequirementCreateWithString(CFSTR("entitlement[\"com.apple.security.app-sandbox\"] exists"), kSecCSDefaultFlags, &sandboxRequirement);
    if (!sandboxRequirement) {
        return NO;
    }
    OSStatus codeCheckResult = SecStaticCodeCheckValidityWithErrors(staticCode, kSecCSBasicValidateOnly, sandboxRequirement, NULL);
    if (codeCheckResult == errSecSuccess) {
        isSandboxed = YES;
    }
    CFRelease(staticCode);
    CFRelease(sandboxRequirement);
    return isSandboxed;
}

File: core/dylib/src/main/java/ch/cyberduck/core/sparkle/Sandbox.java
package ch.cyberduck.core.sparkle;

/*
 * Copyright (c) 2002-2014 David Kocher. All rights reserved.
 * http://cyberduck.io/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to:
 * feedback@cyberduck.io
 */

import ch.cyberduck.core.library.Native;

import org.apache.log4j.Logger;

public final class Sandbox {
    private static final Logger log = Logger.getLogger(Sandbox.class);

    static {
        Native.load("core");
    }

    public static Sandbox get() {
        if(log.isDebugEnabled()) {
            log.debug("Obtain sandbox environment with requirement com.apple.security.app-sandbox");
        }
        return new Sandbox();
    }

    public native boolean isSandboxed();
}
